evm/vm/base.py

```diff
@@ -235,6 +236,13 @@ def pack_block(self, block, *args, **kwargs):
 
         return block
 
+    @contextlib.contextmanager
+    def state_in_temp_block(self):
+        header = self.block.header
+        temp_block = self.generate_block_from_parent_header_and_coinbase(header, header.coinbase)
+        prev_headers = [self.block.header] + self.prev_headers
+        yield self.get_state(block_header=temp_block.header, prev_headers=prev_headers)
```

Comment:
Do I need to do more cleanup here to undo changes that happen in the `with` context? I don't understand all the internals, yet.

Reply:
After another look, maybe it should be something like:

```py
state = self.get_state(block_header=temp_block.header, prev_headers=prev_headers)
snapshot = state.snapshot()
yield state
state.revert(snapshot)
```

Reply:
This looks closer to correct.  Assuming the context manager API for state is gone, this should be fine.  If it's not, then you need to nest the `yield` in it's own `with` block.

Reply:
I see a `state_db` context manager, but it's invoked several times inside the various VM execute transactions, with no way to pass a `read_only` flag. On a first-pass read-through of the code, `state.snapshot()` and `state.revert()` should reverse any of those changes, though.

Reply:
I believe you are correct.  I think @hwwhww was looking into removing the context manager API because it no longer makes sense with some of the recent changes, but that may not have happened yet.